### PR TITLE
cmake: honor custom `CMAKE_UNITY_BUILD_BATCH_SIZE`

### PR DESCRIPTION
**Pull request by vszakats**
This value tells how many sources files to bundle in a single "unity"
compilation unit.

The CMake default is 8 sources, curl's CMake set this to 0, meaning
to bundle all sources into a single unit.

This patch makes it possible to override the 0 value, and potentially
optimize the build process further by better utilizing multiple cores
in conjunction with `make -jN`.

The number of sources in lib is 172 at the time of writing this. For
a 12-core CPU, this can give a job for them all:
`-DCMAKE_UNITY_BUILD_BATCH_SIZE=15`

(Compile time may be affected by a bunch of other factors.)

### CODE REVIEW

**Comment by bagder** [comment]
For some reason this feature breaks the build for me in ways it does not for "normal" unity build:

~~~
$ cmake -B build -DCMAKE_UNITY_BUILD=ON -DCMAKE_UNITY_BUILD_BATCH_SIZE=15
$ cd build
$make -sj
...

/home/dast/src/curl/src/tool_getparam.c: In function ‘getparameter’:
/home/dast/src/curl/src/tool_getparam.c:2409:11: error: implicit declaration of function ‘msnprintf’; did you mean ‘vsnprintf’? [-Wimplicit-function-declaration]
 2409 |           msnprintf(buffer, sizeof(buffer), "%" CURL_FORMAT_CURL_OFF_T "-",
      |           ^~~~~~~~~
      |           vsnprintf
/home/dast/src/curl/src/tool_getparam.c:2409:11: warning: nested extern declaration of ‘msnprintf’ [-Wnested-externs]
In file included from /home/dast/src/curl/build/src/CMakeFiles/curl.dir/Unity/unity_1_c.c:28:
/home/dast/src/curl/src/tool_ipfs.c: In function ‘ipfs_gateway’:
/home/dast/src/curl/src/tool_ipfs.c:83:19: error: implicit declaration of function ‘aprintf’; did you mean ‘dprintf’? [-Wimplicit-function-declaration]
   83 |       ipfs_path = aprintf("%s/.ipfs/", home);
      |                   ^~~~~~~
      |                   dprintf
/home/dast/src/curl/src/tool_ipfs.c:83:19: warning: nested extern declaration of ‘aprintf’ [-Wnested-externs]
/home/dast/src/curl/src/tool_ipfs.c:83:17: error: assignment to ‘char *’ from ‘int’ makes pointer from integer without a cast [-Wint-conversion]
   83 |       ipfs_path = aprintf("%s/.ipfs/", home);
      |                 ^
/home/dast/src/curl/src/tool_ipfs.c:90:30: error: assignment to ‘char *’ from ‘int’ makes pointer from integer without a cast [-Wint-conversion]
   90 |   gateway_composed_file_path = aprintf("%sgateway", ipfs_path);
      |                              ^
/home/dast/src/curl/src/tool_ipfs.c: In function ‘ipfs_url_rewrite’:
/home/dast/src/curl/src/tool_ipfs.c:240:14: error: assignment to ‘char *’ from ‘int’ makes pointer from integer without a cast [-Wint-conversion]
  240 |   pathbuffer = aprintf("%s%s/%s%s", gwpath, protocol, cid,
      |              ^
In file included from /home/dast/src/curl/build/src/CMakeFiles/curl.dir/Unity/unity_1_c.c:43:
/home/dast/src/curl/src/tool_operate.c: In function ‘single_transfer’:
/home/dast/src/curl/src/tool_operate.c:986:20: error: assignment to ‘char *’ from ‘int’ makes pointer from integer without a cast [-Wint-conversion]
  986 |             header = aprintf("If-None-Match: %s", etag_from_file);
      |                    ^
/home/dast/src/curl/src/tool_operate.c:990:20: error: assignment to ‘char *’ from ‘int’ makes pointer from integer without a cast [-Wint-conversion]
  990 |             header = aprintf("If-None-Match: \"\"");
      |                    ^
/home/dast/src/curl/src/tool_operate.c:1197:23: error: initialization of ‘char *’ from ‘int’ makes pointer from integer without a cast [-Wint-conversion]
 1197 |             char *d = aprintf("%s/%s", config->output_dir, per->outfile);
      |                       ^~~~~~~
In file included from /home/dast/src/curl/build/src/CMakeFiles/curl.dir/Unity/unity_1_c.c:46:
/home/dast/src/curl/src/tool_operhlp.c: In function ‘add_file_name_to_url’:
/home/dast/src/curl/src/tool_operhlp.c:144:19: error: assignment to ‘char *’ from ‘int’ makes pointer from integer without a cast [-Wint-conversion]
  144 |           newpath = aprintf("%s%s", path, encfile);
      |                   ^
/home/dast/src/curl/src/tool_operhlp.c:147:19: error: assignment to ‘char *’ from ‘int’ makes pointer from integer without a cast [-Wint-conversion]
  147 |           newpath = aprintf("%s/%s", path, encfile);
      |                   ^
make[2]: *** [src/CMakeFiles/curl.dir/build.make:115: src/CMakeFiles/curl.dir/Unity/unity_1_c.c.o] Error 1

**Comment by vszakats** [comment]
Right, this shouldn't of course happen, but with this setting there is any number of new mash-up combinations, and it isn't without side-effects. I could replicate, and having a look.

**Comment by bagder** [comment]
I don't think you need to spend a lot of energy on fixing that if it turns out cumbersome. It's such a niche thing.

**Comment by vszakats** [comment]
I think it's useful because issues like this may pop up in other places in the future. Even in UNITY_BATCH=0.

In this case it was the common one where a header guarded against multiple inclusion has two operating modes depending on a macro set before including it: `curlx.h` and `ENABLE_CURLX_PRINTF`.

With BATCH=15 a source included it without setting `ENABLE_CURLX_PRINTF`, so the next source trying to get these redefines did not get them!

Basically the same as the `libssh.h` issue fixed yesterday.
